### PR TITLE
[AIR] Add Torch image example

### PR DESCRIPTION
**Pull request by bveeramani**
**IMPORTANT**: This PR is stacked on:
- [x] #25190 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,941 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2afbd6e7",
+   "metadata": {},
+   "source": [
+    "* Install the [Ray AI Runtime](https://docs.ray.io/en/latest/ray-air/getting-started.html). You'll need Ray 1.13 later to run this example.\n",
+    "\n",
+    "```\n",
+    "pip instsall 'ray[data,tune]'\n",
```
Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:37]:
```suggestion
    "pip install 'ray[data,tune]'\n",
```
```suggestion
    "pip instsall 'ray[data,tune]'\n",
```

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,941 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ad65d385",
+   "metadata": {},
+   "source": [
+    "This tutorial illustrates how to train an image classifier using the [Ray AI Runtime](https://docs.ray.io/en/latest/ray-air/getting-started.html).\n",
+    "\n",
+    "You should be familiar with [PyTorch](https://pytorch.org/) before starting the tutorial. If you need a refresher, read PyTorch's [training a classifier](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) tutorial."
```
Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:18]: Nice! I love having explicit prerequisites

Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:18]: Nice! I love having explicit prerequisites

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,941 @@
...
+    "pip install torch torchvision\n",
+    "```\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "af1bfc48",
+   "metadata": {},
+   "source": [
+    "## Load and normalize CIFAR-10"
```
Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:52]: Can you add some text explaining what's going on here?

Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:52]: Can you add some text explaining what's going on here?

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,941 @@
...
+     "output_type": "stream",
+     "text": [
+      "\u001b[2m\u001b[36m(_execute_read_task pid=14274)\u001b[0m Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to ./data/cifar-10-python.tar.gz\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "0.0%\u001b[36m(_execute_read_task pid=14274)\u001b[0m \n",
```
Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:81]: This progress bar looks really bad-- @xwjiang2010 @clarkzinzow this was the same problem that spotify ran into right?

Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:81]: This progress bar looks really bad-- @xwjiang2010 @clarkzinzow this was the same problem that spotify ran into right?

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,941 @@
...
+      "\u001b[2m\u001b[36m(BaseWorkerMixin pid=14522)\u001b[0m [2, 10000] loss: 1.341\n",
+      "\u001b[2m\u001b[36m(BaseWorkerMixin pid=14521)\u001b[0m [2, 12000] loss: 1.262\n",
+      "\u001b[2m\u001b[36m(BaseWorkerMixin pid=14522)\u001b[0m [2, 12000] loss: 1.265\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "2022-05-09 14:06:04,518\tERROR checkpoint_manager.py:189 -- Result dict has no key: training_iteration. checkpoint_score_attr must be set to a key of the result dict. Valid keys are ['trial_id', 'experiment_id', 'date', 'timestamp', 'pid', 'hostname', 'node_ip', 'config', 'done']\n"
```
Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:756]: cc @xwjiang2010 this is the same bug that Spotify ran into

Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:756]: cc @xwjiang2010 this is the same bug that Spotify ran into

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,941 @@
...
+    }
+   ],
+   "source": [
+    "\n",
+    "import ray\n",
+    "from ray.data.datasource import SimpleTorchDatasource\n",
+    "import torchvision\n",
+    "import torchvision.transforms as transforms\n",
+    "\n",
+    "transform = transforms.Compose(\n",
```
Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:520]: Should these be implemented as a `TorchVisionTransformPreprocessor`?

Comment (amogkam) [inline python/ray/ml/examples/pytorch/torch_image_example.ipynb:520]: Should these be implemented as a `TorchVisionTransformPreprocessor`?

**Review by amogkam — COMMENTED** [review]
Looks good overall! I'm assuming more text/description will be coming in later?

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb (hunk 5 above)
Comment (xwjiang2010) [reply python/ray/ml/examples/pytorch/torch_image_example.ipynb:756]: ah got it. let me take a look!

Comment (xwjiang2010) [reply python/ray/ml/examples/pytorch/torch_image_example.ipynb:756]: fyi, https://github.com/ray-project/ray/issues/24593

File: python/ray/ml/examples/pytorch/torch_image_example.ipynb (hunk 6 above)
Comment (bveeramani) [reply python/ray/ml/examples/pytorch/torch_image_example.ipynb]: That's not currently a thing, right?

**Review by richardliaw — CHANGES_REQUESTED** [review]

File: doc/source/ray-air/examples/torch_image_example.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
+---
+jupytext:
+    text_representation:
+        extension: .md
+        format_name: myst
+kernelspec:
+    display_name: Python 3
+    language: python
+    name: python3
+---
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:10]: I was planning on converting `torch_image_example.md` to a Jupyter notebook after review. IMO this tutorial is easier to understand with output cells.

File: doc/source/ray-air/examples/torch_image_example.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+
+# Training a Torch Classifier
+
+This tutorial demonstrates how to train an image classifier using the [Ray AI Runtime](https://docs.ray.io/en/latest/ray-air/getting-started.html) (AIR).
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md]: Fixed

File: doc/source/ray-air/examples/torch_image_example.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
+train_dataset: ray.data.Dataset = ray.data.read_datasource(SimpleTorchDatasource(), dataset_factory=train_dataset_factory)
+test_dataset: ray.data.Dataset = ray.data.read_datasource(SimpleTorchDatasource(), dataset_factory=test_dataset_factory)
+
+train_dataset
+```
+
+Note that [`SimpleTorchDatasource`](https://docs.ray.io/en/master/data/package-ref.html#ray.data.datasource.SimpleTorchDatasource) loads all data into memory, so you shouldn't use it with larger datasets.
+
+Next, let's represent our data using pandas dataframes instead of tuples. This lets us call methods like [`Dataset.to_torch`](https://docs.ray.io/en/master/data/package-ref.html#ray.data.Dataset.to_torch) later in the tutorial.
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md]: Fixed

File: doc/source/ray-air/examples/torch_image_example.md [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+This tutorial demonstrates how to train an image classifier using the [Ray AI Runtime](https://docs.ray.io/en/latest/ray-air/getting-started.html) (AIR).
+
+You should be familiar with [PyTorch](https://pytorch.org/) before starting the tutorial. If you need a refresher, read PyTorch's [training a classifier](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) tutorial.
+
+## Before you begin
+
+* Install the [Ray AI Runtime](https://docs.ray.io/en/latest/ray-air/getting-started.html). You'll need Ray 1.13 later to run this example.
+
+```{code-cell} python3
+!pip install 'ray[data,serve,tune]'
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:24]: Ah, didn't know that. Fixed

File: doc/source/ray-air/examples/torch_image_example.md [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+
+# Training a Torch Classifier
+
+This tutorial demonstrates how to train an image classifier using the [Ray AI Runtime](https://docs.ray.io/en/latest/ray-air/getting-started.html) (AIR).
+
+You should be familiar with [PyTorch](https://pytorch.org/) before starting the tutorial. If you need a refresher, read PyTorch's [training a classifier](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) tutorial.
+
+## Before you begin
+
+* Install the [Ray AI Runtime](https://docs.ray.io/en/latest/ray-air/getting-started.html). You'll need Ray 1.13 later to run this example.
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md]: Fixed

File: doc/source/ray-air/examples/torch_image_example.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:10]: Actually, is there a way to show cell outputs with MyST and the `literalinclude` directive?

File: doc/source/ray-air/examples/torch_image_example.md [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_loop_per_worker,
+    train_loop_config={"batch_size": 2},
+    datasets={"train": train_dataset},
+    scaling_config={"num_workers": 2}
+)
+result = trainer.fit()
+latest_checkpoint = result.checkpoint
+```
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:177]: I've added a small paragraph. Lemme know what you think.

Anecdotally, it's difficult to create a Ray Cluster and scale out, so I'm not sure if we should attempt to explain that in this "Hello World" example. It might be better to just leave a sentence like:

> To scale your training script, read ...

I feel like to appropriately explain scaling out, we'd need to add a lot of information that's out of scope for this example.

File: doc/source/ray-air/examples/torch_image_example.md [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+```{code-cell} python3
+from ray.ml.predictors.integrations.torch import TorchPredictor
+from ray.ml.batch_predictor import BatchPredictor
+
+batch_predictor = BatchPredictor.from_checkpoint(
+    checkpoint=latest_checkpoint,
+    predictor_cls=TorchPredictor,
+    model=Net(),
+)
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:201]: Fixed

File: doc/source/ray-air/examples/torch_image_example.md [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+    
+
+train_dataset = train_dataset.map_batches(convert_batch_to_pandas)
+test_dataset = test_dataset.map_batches(convert_batch_to_pandas)
+
+train_dataset
+```
+
+## Train a convolutional neural network
+
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:87]: What did you have in mind? Like, a description of the model definition, or a high-level description of the section?

[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/ray-air/examples/torch_image_example.md:87]: High level description of the section.

**Review by richardliaw — CHANGES_REQUESTED** [review]
~Ah sorry, forgot to leave one more comment. is this file tested in CI? If not, can we convert it to a notebook/python script form so that we can test in CI?~

nvm saw the above thread

File: doc/source/ray-air/examples/torch_image_example.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/ray-air/examples/torch_image_example.md:10]: You can run it locally and copy the outputs into the file I guess?

Otherwise fine to use the notebook format.

File: doc/source/ray-air/examples/torch_image_example.md [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+)
+
+outputs: ray.data.Dataset = batch_predictor.predict(
+    data=test_dataset, feature_columns=["image"], unsqueeze=False
+)
+
+outputs.show(1)
+```
+
+```{code-cell} python3
```
Comment (richardliaw) [inline doc/source/ray-air/examples/torch_image_example.md:212]: Can you explain what you're doing in this cell?

File: doc/source/ray-air/examples/torch_image_example.md [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+    return df[["prediction"]]
+
+predictions = outputs.map_batches(
+    convert_logits_to_classes, batch_format="pandas"
+)
+
+predictions.show(1)
+```
+
+```{code-cell} python3
```
Comment (richardliaw) [inline doc/source/ray-air/examples/torch_image_example.md:227]: Can you explain what you're doing in this cell?

File: doc/source/ray-air/examples/torch_image_example.md [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+def calculate_prediction_scores(df):
+    df["correct"] = df["prediction"] == df["label"]
+    return df[["prediction", "label", "correct"]]
+
+scores = test_dataset.zip(predictions).map_batches(calculate_prediction_scores)
+
+scores.show(1)
+```
+
+```{code-cell} python3
```
Comment (richardliaw) [inline doc/source/ray-air/examples/torch_image_example.md:237]: Can you explain what you're doing in this cell?

File: doc/source/ray-air/examples/torch_image_example.md [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+scores.sum(on="correct") / scores.count()
+```
+
+## Deploy the network and make a prediction
+
+```{code-cell} python3
```
Comment (richardliaw) [inline doc/source/ray-air/examples/torch_image_example.md:243]: Can you explain what you're doing in this cell?

**Review by richardliaw — COMMENTED** [review]
Just a couple more explanations to add!

File: doc/source/ray-air/examples/torch_image_example.md (hunk 17 above)
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:240]: Fixed

File: doc/source/ray-air/examples/torch_image_example.md (hunk 18 above)
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:249]: Fixed

File: doc/source/ray-air/examples/torch_image_example.md (hunk 15 above)
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:215]: Fixed

File: doc/source/ray-air/examples/torch_image_example.md (hunk 16 above)
Comment (bveeramani) [reply doc/source/ray-air/examples/torch_image_example.md:236]: Fixed

**Review by richardliaw — APPROVED** [review]
looks good! one nit.

**Comment by richardliaw** [comment]
Also you will convert to a runnable format right?

**Comment by bveeramani** [comment]
> Also you will convert to a runnable format right?

Yeah. I'm going to convert it to a Jupyter notebook

**Comment by richardliaw** [comment]
Can you make sure `balaji` doesn't show up on the stdout in the notebook?

**Comment by richardliaw** [comment]
There's a couple places with `/Users/balaji/`

**Comment by bveeramani** [comment]
> There's a couple places with `/Users/balaji/`

Yeah, was wondering if that'd be an issue. Will fix.

**Comment by zhe-thoughts** [comment]
@jianoaix @clarkzinzow @pcmoritz : We are also trying to add an image example for Datasets. How is that example different from this one?

**Comment by clarkzinzow** [comment]
@zhe-thoughts This example uses AIR, while the upcoming imagery example for Datasets GA will not use AIR.

**Comment by pcmoritz** [comment]
Also I hope the dataset example will do some more preprocessing with the images than this example, e.g. resize them all to the same size :)
